### ISSUE TITLE
LinkLoader example

### ISSUE THREAD

**Issue by daviddavo** [pyg-team/pytorch_geometric#7817]
### 📚 Describe the documentation issue

I want to use the LinkLoader class, but it asks for a `link_sampler` which I don't have. I haven't found an implementation of a LinkSampler online, and I don't know where to start.

I think it would be far easier for me to just ditch using PyG's classes and use a generic Pytorch DataLoader with the edge_index. In fact, this is what official examples like [LightGCN example](https://github.com/pyg-team/pytorch_geometric/blob/master/examples/lightgcn.py) do:

https://github.com/pyg-team/pytorch_geometric/blob/02cc18d9d6841ecda4eb61eebb48b86f1aaae477/examples/lightgcn.py#L20-L26

### Suggest a potential alternative/fix

It would be easier to understand if the official examples used Pytorch geometric's classes for sampling and data loading.

**Comment by rusty1s** [comment]
Yes, you can also use the `DataLoader` if you are not interested in mini-batch GNN execution. Besides that, there is no need to pass a custom sampler if you are using `LinkNeighborLoader` instead.

**Comment by daviddavo** [comment]
I recently realized that [the original LightGCN implementation](https://github.com/gusye1234/LightGCN-PyTorch/blob/master/code/Procedure.py) used BPR sampling. (For every _user_ it returns a _positive_ sample and a _negative_ sample). An epoch is just iterating through every user.

Why was it changed in PyG's example?

**Comment by akihironitta** [comment]
> For every user it returns a positive sample and a negative sample

I'm not sure if I understand your point, but as far as I understand, both of them do basically the same thing:
- [PyG's LightGCN example](https://github.com/pyg-team/pytorch_geometric/blob/60ea78bc2fbb07bf10080889ab27265e6da1ee31/examples/lightgcn.py#L40-L50)
- https://github.com/gusye1234/LightGCN-PyTorch/blob/947ca2b3b1d2d3545b114145710cb06c4e57b3d2/code/Procedure.py#L33-L49

**Comment by daviddavo** [comment]
The difference is that PyG's example creates a negative sample for every edge. A user that interacted 8 times with an item, will get 8 positive "samples" (they are not samples, they are all the data available) and 8 negative samples. Meanwhile, in paperswithcode implementation, they sample one positive and one negative per user.

They also iterate through users instead of through links.

**Comment by rusty1s** [comment]
I think this still holds in PyG. For every link we basically sample x many negatives (for which you can specify x=1). The `LinkNeighborLoader` samples links, not entities.

**Comment by daviddavo** [comment]
As I understand it:
- In the original paper implementation, an **epoch** is passing one pos/neg sample for every user in the dataset. It's a kind of user-wise sampling. [Sampling source code](https://github.com/gusye1234/LightGCN-PyTorch/blob/947ca2b3b1d2d3545b114145710cb06c4e57b3d2/code/utils.py#L78-L92)
- In PyG's example, an **epoch** is iterating through every rating of the dataset. It's a kind of rating-wise sampling.

**Therefore, the number of epochs on both models are not comparable.**

Furthermore, as each user might have a different number of ratings:
- In the original paper implementation, after N epochs, each user would have seen N samples.
- In PyG's example, after N epochs a user with 20 ratings would have seen 20N samples, while a user with just 2 rates would see 2N samples. 

I don't know which one of these approaches would be better. Nevertheless, they might produce different results, and someone replicating the original experiment following this library might be disappointed (or positively surprised)

**Comment by rusty1s** [comment]
Ok, understood. This isn't too hard to "fix". We just need to iterate over users, and then query all the `edge_label_index` entries for sampled users. Let me take a look at this :)

**Comment by daviddavo** [comment]
FYI, I looked on other libraries, and Microsft's recommenders seems to do "user-wise" sampling as well:

```python
    def train_loader(self, batch_size):
        """Sample train data every batch. One positive item and one negative item sampled for each user.

        Args:
            batch_size (int): Batch size of users.

        Returns:
            numpy.ndarray, numpy.ndarray, numpy.ndarray:
            - Sampled users.
            - Sampled positive items.
            - Sampled negative items.
        """
```

https://github.com/recommenders-team/recommenders/blob/2aca74f6bb1f77d41bbb221e0137d5284ccc24f5/recommenders/models/deeprec/DataModel/ImplicitCF.py#L194C5-L227C73
